Team417_CENTERSTAGE/src/main/java/org/firstinspires/ftc/team417_CENTERSTAGE/roadrunner/MecanumDrive.java

```diff
@@ -0,0 +1,579 @@
+import com.acmerobotics.roadrunner.ftc.LynxFirmware;
+import com.acmerobotics.roadrunner.ftc.OverflowEncoder;
+import com.acmerobotics.roadrunner.ftc.PositionVelocityPair;
+import com.acmerobotics.roadrunner.ftc.RawEncoder;
+import com.qualcomm.hardware.lynx.LynxModule;
+import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
+import com.qualcomm.robotcore.hardware.DcMotor;
+import com.qualcomm.robotcore.hardware.DcMotorEx;
+import com.qualcomm.robotcore.hardware.HardwareMap;
+import com.qualcomm.robotcore.hardware.IMU;
+import com.qualcomm.robotcore.hardware.VoltageSensor;
+import com.qualcomm.robotcore.util.ElapsedTime;
+
+import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
+import org.firstinspires.ftc.team417_CENTERSTAGE.apriltags.AprilTagLatencyCompensation;
+import org.firstinspires.ftc.team417_CENTERSTAGE.apriltags.AprilTagPoseEstimator;
+import org.firstinspires.ftc.team417_CENTERSTAGE.utilityclasses.TwistWithTimestamp;
+import org.firstinspires.inspection.InspectionState;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+
+@Config
+public final class MecanumDrive {
+    // For the April Tag latency calculation (added by Hank)
+    public ElapsedTime clock = new ElapsedTime();
+
+    // To keep a record of twists to be used by April Tag latency compensation (added by Hank)
+    public ArrayList<TwistWithTimestamp> twistList;
+
+    // To detect April Tags to correct drift (added by Hank)
+    public AprilTagPoseEstimator myAprilTagPoseEstimator;
+
+    // To detect the motion of the motors (added by Hank)
+    public DcMotorEx[] motors;
+
+    // Whether or not to use April Tags (added by Hank)
+    public final static boolean USE_APRIL_TAGS = false;
+
+    public static String getBotName() {
+        InspectionState inspection=new InspectionState();
+        inspection.initializeLocal();
+        Log.d("roadrunner", String.format("Device name:" + inspection.deviceName));
+        return inspection.deviceName;
+    }
+    public static boolean isDevBot = getBotName().equals("DevBot");
+
+    public static class Params {
+        public double inPerTick;
+        public double lateralInPerTick;
+        public double trackWidthTicks;
+
+        public double kS;
+        public double kV;
+        public double kA;
+
+        public double maxWheelVel;
+        public double minProfileAccel;
+        public double maxProfileAccel;
+
+        public double maxAngVel;
+        public double maxAngAccel;
+
+        public double axialGain;
+        public double lateralGain;
+        public double headingGain;
+
+        public double axialVelGain;
+        public double lateralVelGain;
+        public double headingVelGain;
+
+        Params() {
+            if (isDevBot) {
+                // drive model parameters
+                inPerTick = 0.0225669957686882; // 96.0 / 4254.0;
+                lateralInPerTick = 0.020179372197309417; // 49.5 / 2453
```

Comment:
These comments are not very useful since I don't know what any of the numbers mean. Most of the numbers in this file need an explanation as to how you determined them.